### PR TITLE
[change] Clean up trainer interface, clean up GhostTrainer stats

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

This PR moves most private methods out of the Trainer class and into RLTrainer, making it more of an interface. This also removed some methods in the Ghost Trainer. As part of this change, it was necessary to move the console printing out of the Ghost Trainer and into the ConsoleWriter. @andrewcoh I need to make sure that I'm writing the stats in the right place, and that the ELO reported where it is won't cause issues.

Also moved reward reporting from the AgentProcessor to the Trainer, so that the GhostTrainer's cumulative reward is (correctly) only one of the learning teams. This doesn't affect single-team training.

There is some special casing and property passing in the ConsoleWriter that is a bit hacky - open to ideas on how to solve that. 

Also fixes a bug where hyperparameters were printed twice for ghost trainer. 

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/ghost/trainer.py [hunk 1]
```diff
@@ -203,6 +190,11 @@ def add_policy(self, name_behavior_id: str, policy: TFPolicy) -> None:
             self.trainer.add_policy(name_behavior_id, policy)
             self._save_snapshot(policy)  # Need to save after trainer initializes policy
             self.learning_behavior_name = name_behavior_id
+            behavior_id_parsed = BehaviorIdentifiers.from_name_behavior_id(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/ghost/trainer.py]: Should we send the behavior ids as a named tuple to all trainers/ghost trainer instead of parsing this in every advance step?

File: ml-agents/mlagents/trainers/stats.py [hunk 2]
```diff
@@ -77,6 +79,9 @@ def write_stats(
 class ConsoleWriter(StatsWriter):
     def __init__(self):
         self.training_start_time = time.time()
+        # If self-play, we want to print ELO instead of reward
```
[the remark this answers is not in the document]
Comment (ervteng) [reply ml-agents/mlagents/trainers/stats.py]: It's in addition to - I didn't change how the ghosted agent adds the reward (it's still done by the AgentProcessor/Manager). 

Currently the trainer's "Extrinsic Reward" log (under Policy in TB) is what we want. We can do a hack in the ConsoleWriter to output Extrinsic Reward instead of Cumulative Reward (meaningless) as the reward if using self-play

File: ml-agents/mlagents/trainers/ghost/trainer.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/ghost/trainer.py]: This is being parsed once right? Just at add policy. 

But you're right - maybe we should parse the behavior ID once in the trainer controller rather than inside the trainer, and introduce a new class (BehaviorIdentifier, with properties behavior_id and team_id).

File: ml-agents/mlagents/trainers/stats.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/stats.py]: How would you feel about that? It could resolve this https://jira.unity3d.com/browse/MLA-760 in part.

File: ml-agents/mlagents/trainers/ghost/trainer.py (hunk 1 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/ghost/trainer.py]: The object the parsing returns is a named tuple.  It would probably be good to pass that around instead.  I can make this in a separate PR since I was considering it for multiagent training too.
